README.md
# televize #
Script to play Czech Television (Česká televize) streams in custom player.

### Installation ###
Directly from pip (preferred)
```sh
pip install televize
```
or
```sh
pip install https://github.com/ziima/televize/archive/master.zip
```
or
```sh
python setup.py install
```

### Dependencies ###
 * python 3.4 or 3.5
 * Any player which supports internet playlists. mpv is used by default.
 * Other dependencies are listed in [requirements.txt](requirements.txt).

### Usage ###
```sh
televize live 24  # CT24
televize live D  # CT:D
televize live sport  # CT-sport
televize ivysilani http://www.ceskatelevize.cz/ivysilani/11276561613-kosmo/
televize ivysilani http://www.ceskatelevize.cz/porady/11276561613-kosmo/215512121020005-triumf
```

#### macOS ####

To open stream in a native player on macOS (e.g. VLC), use `--player` command `open -a APP --args`.
For example:

```sh
televize --player "open -a VLC --args" live 24
```

Based on Perl scripts `televize` and `ctstream` from Petr Písař (http://xpisar.wz.cz).


Changelog.md
## 0.4 ##
 * Fix pypi upload

## 0.3 ##
 * Add support for "pořady" URLs.
 * Fix live streams after CT update in April.

## 0.2 ##
 * Add `--quality` argument to select stream quality.
 * Better exception handling.


tests/utils.py
"""
Utilities for tests
"""
import os
from io import BytesIO

from requests import Response


def get_path(filepath, path):
    """
    Returns full path relative to the file.
    """
    return os.path.join(os.path.dirname(filepath), path)


def make_response(content):
    """
    Returns `requests.Response` object with defined `content`.
    """
    response = Response()
    response.raw = BytesIO(content)
    return response


tests/test_live.py
"""
Test live streaming
"""
import unittest
from subprocess import PIPE
from unittest.mock import Mock, call, patch, sentinel

from m3u8.model import Playlist

from televize import play_live

from .utils import get_path, make_response


class TestPlayLive(unittest.TestCase):
    """Test `play_live` function"""
    def setUp(self):
        patcher = patch('televize.requests')
        self.addCleanup(patcher.stop)
        self.requests_mock = patcher.start()

        popen_patcher = patch('televize.Popen')
        self.addCleanup(popen_patcher.stop)
        self.popen_mock = popen_patcher.start()

    def test_play_live(self):
        self.requests_mock.get.side_effect = [
            make_response(open(get_path(__file__, "data/play_live/playlist_1.m3u"), mode='rb').read()),
            Mock(content=sentinel.data_1),
            Mock(content=sentinel.data_2),
            Mock(content=sentinel.data_3),
            make_response(open(get_path(__file__, "data/play_live/playlist_2.m3u"), mode='rb').read()),
            Mock(content=sentinel.data_4),
        ]
        sleep_mock = Mock()
        playlist_uri = 'http://80.188.78.151:80/atip/fd2eccaa99022586e14694df91068915/1449324471384/' \
                       '3616440c710a1d7e3f54761a6d940c64/2402-tv-pc/1502.m3u8'
        playlist = Playlist(playlist_uri, {'bandwidth': 500000}, None, None)

        play_live(playlist, 'my_player "Custom Argument" --', _sleep=sleep_mock)

        calls = [
            call.get('http://80.188.78.151:80/atip/fd2eccaa99022586e14694df91068915/1449324471384/'
                     '3616440c710a1d7e3f54761a6d940c64/2402-tv-pc/1502.m3u8'),
            call.get('http://80.188.78.151:80/atip/fd2eccaa99022586e14694df91068915/1449324471384/'
                     '3616440c710a1d7e3f54761a6d940c64/2402-tv-pc/1502/58877187.ts'),
            call.get('http://80.188.78.151:80/atip/fd2eccaa99022586e14694df91068915/1449324471384/'
                     '3616440c710a1d7e3f54761a6d940c64/2402-tv-pc/1502/58877188.ts'),
            call.get('http://80.188.78.151:80/atip/fd2eccaa99022586e14694df91068915/1449324471384/'
                     '3616440c710a1d7e3f54761a6d940c64/2402-tv-pc/1502/58877189.ts'),
            call.get('http://80.188.78.151:80/atip/fd2eccaa99022586e14694df91068915/1449324471384/'
                     '3616440c710a1d7e3f54761a6d940c64/2402-tv-pc/1502.m3u8'),
            call.get('http://80.188.78.151:80/atip/fd2eccaa99022586e14694df91068915/1449324471384/'
                     '3616440c710a1d7e3f54761a6d940c64/2402-tv-pc/1502/58877190.ts'),
        ]
        self.assertEqual(self.requests_mock.mock_calls, calls)
        self.assertEqual(sleep_mock.mock_calls, [call(8.0)])
        popen_calls = [call(['my_player', 'Custom Argument', '--'], stdin=PIPE),
                       call().stdin.write(sentinel.data_1),
                       call().stdin.write(sentinel.data_2),
                       call().stdin.write(sentinel.data_3),
                       call().stdin.write(sentinel.data_4)]
        self.assertEqual(self.popen_mock.mock_calls, popen_calls)


tests/test_play.py
"""
Test live streaming
"""
import unittest
from unittest.mock import call, patch

import responses
from m3u8.model import Playlist

from televize import play, run_player

from .utils import get_path


class TestRunPlayer(unittest.TestCase):
    """Test `run_player` function"""
    def setUp(self):
        call_patcher = patch('televize.subprocess.call')
        self.addCleanup(call_patcher.stop)
        self.call_mock = call_patcher.start()

    def test_run_player_live(self):
        playlist_uri = 'http://example.cz/path/playlist.m3u8'
        playlist = Playlist(playlist_uri, {'bandwidth': 500000}, None, None)

        run_player(playlist, 'my_player "Custom Argument"')

        self.assertEqual(self.call_mock.mock_calls, [call(['my_player', 'Custom Argument', playlist_uri])])


class TestPlay(unittest.TestCase):
    """Test `play` function"""

    def setUp(self):
        call_patcher = patch('televize.subprocess.call')
        self.addCleanup(call_patcher.stop)
        self.call_mock = call_patcher.start()

    def test_play_live(self):
        options = {'live': True, '<channel>': '24', '--player': 'mpv'}
        playlist_url = 'http://www.ceskatelevize.cz/ivysilani/client-playlist/?key=df365c9c2ea8b36f76dfa29e3b16d245'

        with responses.RequestsMock() as rsps:
            rsps.add(responses.POST, 'http://www.ceskatelevize.cz/ivysilani/ajax/get-client-playlist',
                     json={'url': playlist_url}),
            rsps.add(responses.GET, 'http://www.ceskatelevize.cz/ivysilani/client-playlist/',
                     body=open(get_path(__file__, 'data/play_live/client_playlist.json')).read())
            rsps.add(responses.GET, 'http://80.188.65.18:80/cdn/uri/get/',
                     body=open(get_path(__file__, 'data/play_live/stream_playlist.m3u')).read())
            play(options)

        stream_url = ('http://80.188.78.151:80/atip/fd2eccaa99022586e14694df91068915/1449324471384/'
                      '3616440c710a1d7e3f54761a6d940c64/2402-tv-pc/1502.m3u8')
        self.assertEqual(self.call_mock.mock_calls, [call(['mpv', stream_url])])

    def test_play_live_unknown(self):
        options = {'live': True, '<channel>': 'unknown'}

        with self.assertRaisesRegexp(ValueError, "^Unknown live channel 'unknown'$"):
            play(options)

    def test_play_ivysilani(self):
        options = {'live': False, 'ivysilani': True, '<url>': 'http://www.ceskatelevize.cz/ivysilani/kosmo.html',
                   '--player': 'mpv'}
        playlist_url = 'http://www.ceskatelevize.cz/ivysilani/client-playlist/?key=df365c9c2ea8b36f76dfa29e3b16d245'

        with responses.RequestsMock() as rsps:
            rsps.add(responses.GET, 'http://www.ceskatelevize.cz/ivysilani/kosmo.html',
                     body=open(get_path(__file__, 'data/ivysilani.html'), mode='r').read())
            rsps.add(responses.POST, 'http://www.ceskatelevize.cz/ivysilani/ajax/get-client-playlist',
                     json={'url': playlist_url}),
            rsps.add(responses.GET, 'http://www.ceskatelevize.cz/ivysilani/client-playlist/',
                     body=open(get_path(__file__, 'data/play_live/client_playlist.json')).read())
            rsps.add(responses.GET, 'http://80.188.65.18:80/cdn/uri/get/',
                     body=open(get_path(__file__, 'data/play_live/stream_playlist.m3u')).read())
            play(options)

        stream_url = ('http://80.188.78.151:80/atip/fd2eccaa99022586e14694df91068915/1449324471384/'
                      '3616440c710a1d7e3f54761a6d940c64/2402-tv-pc/1502.m3u8')
        self.assertEqual(self.call_mock.mock_calls, [call(['mpv', stream_url])])


televize.py
#!/usr/bin/env python3
"""
Plays Czech television stream in custom player

Usage: televize.py [options] live <channel>
       televize.py [options] ivysilani <url>
       televize.py -h | --help
       televize.py --version

Subcommands:
  live                 play live channel
  ivysilani            play video from ivysilani archive

Live channels:
  1                    CT1
  2                    CT2
  24                   CT24
  sport                CTsport
  D                    CT:D
  art                  CTart

Options:
  -h, --help           show this help message and exit
  --version            show program's version number and exit
  -q, --quality=QUAL   select stream quality [default: min]. Possible values are integers, 'min' and 'max'.
  -p, --player=PLAYER  player command [default: mpv]
  -d, --debug          print debug messages
"""
import logging
import re
import shlex
import subprocess
import sys
from collections import OrderedDict
from urllib.parse import urljoin, urlsplit

import m3u8
import requests
from docopt import docopt
from lxml import etree

__version__ = '0.4'


PLAYLIST_LINK = 'http://www.ceskatelevize.cz/ivysilani/ajax/get-client-playlist'
CHANNEL_NAMES = OrderedDict((
    ('1', 1),
    ('2', 2),
    ('24', 24),
    ('sport', 4),
    ('D', 5),
    ('art', 6),
))
PLAYLIST_TYPE_CHANNEL = 'channel'
PLAYLIST_TYPE_EPISODE = 'episode'

PORADY_PATH_PATTERN = re.compile('^/porady/[^/]+/(?P<playlist_id>\d+)(-[^/]*)?/?$')


################################################################################
# Playlist functions
def parse_quality(value: str) -> int:
    """Return quality selector parsed from user input value.

    @raises ValueError: If value is not a valid quality selector.
    """
    # Special keywords
    if value == 'min':
        return 0
    elif value == 'max':
        return -1
    try:
        return int(value)
    except ValueError:
        raise ValueError("Quality '{}' is not a valid value.".format(value))


def get_playlist(playlist_id, playlist_type, quality: int):
    """
    Extract the playlist for CT video.

    @param playlist_id: ID of playlist
    @param playlist_type: Type of playlist
    @param quality: Quality selector
    """
    assert playlist_type in (PLAYLIST_TYPE_CHANNEL, PLAYLIST_TYPE_EPISODE)
    # First get the custom client playlist URL
    post_data = {
        'playlist[0][id]': playlist_id,
        'playlist[0][type]': playlist_type,
        'requestUrl': '/ivysilani/',
        'requestSource': "iVysilani",
        'addCommercials': 0,
        'type': "html"
    }
    response = requests.post(PLAYLIST_LINK, post_data, headers={'x-addr': '127.0.0.1'})
    client_playlist = response.json()

    # Get the custom playlist URL to get playlist JSON meta data (including playlist URL)
    response = requests.get(urljoin(PLAYLIST_LINK, client_playlist["url"]))
    playlist_metadata = response.json()
    stream_playlist_url = playlist_metadata['playlist'][0]['streamUrls']['main']

    # Use playlist URL to get the M3U playlist with streams
    response = requests.get(urljoin(PLAYLIST_LINK, stream_playlist_url))
    logging.debug("Variant playlist: %s", response.text)
    playlist_base_url = response.url
    variant_playlist = m3u8.loads(response.text)

    # Select stream based on quality
    playlists = sorted(variant_playlist.playlists, key=lambda p: p.stream_info.bandwidth)
    try:
        playlist = playlists[quality]
    except IndexError:
        raise ValueError("Requested quality {} is not available.".format(quality))
    playlist.base_uri = playlist_base_url
    return playlist


def get_ivysilani_playlist(url, quality: int):
    """
    Extract the playlist for ivysilani page.

    @param url: URL of the web page
    @param quality: Quality selector
    """
    # Porady pages have playlist ID in URL
    split = urlsplit(url)
    match = PORADY_PATH_PATTERN.match(split.path)
    if match:
        playlist_id = match.group('playlist_id')
        return get_playlist(playlist_id, PLAYLIST_TYPE_EPISODE, quality)

    # Try ivysilani URL
    response = requests.get(url)
    page = etree.HTML(response.text)
    play_button = page.find('.//a[@class="programmeToPlaylist"]')
    if not play_button:
        raise ValueError("Can't find playlist on the ivysilani page.")
    item = play_button.get('rel')
    if not item:
        raise ValueError("Can't find playlist on the ivysilani page.")
    return get_playlist(item, PLAYLIST_TYPE_EPISODE, quality)


def get_live_playlist(channel, quality: int):
    """
    Extract the playlist for live CT channel.

    @param channel: Name of the channel
    @param quality: Quality selector
    """
    return get_playlist(CHANNEL_NAMES[channel], PLAYLIST_TYPE_CHANNEL, quality)


################################################################################
def run_player(playlist: m3u8.model.Playlist, player_cmd: str):
    """Run the video player

    @param playlist: Playlist to be played
    @param player_cmd: Additional player arguments
    """
    cmd = shlex.split(player_cmd) + [playlist.absolute_uri]
    logging.debug("Player cmd: %s", cmd)
    subprocess.call(cmd)


def play(options):
    """Play televize.

    @raises ValueError: In case of an invalid options.
    """
    quality = parse_quality(options['--quality'])
    if options['live']:
        if options['<channel>'] not in CHANNEL_NAMES:
            raise ValueError("Unknown live channel '{}'".format(options['<channel>']))
        playlist = get_live_playlist(options['<channel>'], quality)
    else:
        assert options['ivysilani']
        playlist = get_ivysilani_playlist(options['<url>'], quality)
    run_player(playlist, options['--player'])


def main():
    """Main script."""
    options = docopt(__doc__, version=__version__)

    # Set up logging
    if options['--debug']:
        level = logging.DEBUG
    else:
        level = logging.WARNING
    logging.basicConfig(stream=sys.stderr, level=level, format='%(asctime)s %(levelname)s:%(funcName)s: %(message)s')
    logging.getLogger('iso8601').setLevel(logging.WARN)

    try:
        play(options)
    except Exception as error:
        if level == logging.DEBUG:
            logging.exception("An error occured:")
        else:
            logging.warn("An error occured: %s", error)
        exit(1)
    except KeyboardInterrupt:
        # User killed the program, silence the exception
        exit(0)


if __name__ == '__main__':
    main()
